src/main/java/com/synopsys/integration/alert/channel/jira/util/JiraTransitionHelper.java

```diff
@@ -0,0 +1,138 @@
+        StatusDetailsComponent statusDetails = transition.getTo();
+        StatusCategory statusCategory = statusDetails.getStatusCategory();
+        return StringUtils.equals(expectedStatusCategoryKey, statusCategory.getKey());
+    }
+
+    public boolean transitionIssueIfNecessary(String issueKey, JiraIssueConfigValidator.JiraIssueConfig jiraIssueConfig, ItemOperation operation) throws IntegrationException {
+        Optional<String> optionalTransitionKey = determineTransitionKey(operation);
+        if (optionalTransitionKey.isPresent()) {
+            final Optional<String> transitionName = determineTransitionName(operation, jiraIssueConfig);
+            if (transitionName.isPresent()) {
+                boolean shouldAttemptTransition = isTransitionRequired(issueKey, operation);
+                if (shouldAttemptTransition) {
+                    performTransition(issueKey, transitionName.get());
+                    return true;
+                } else {
+                    logger.debug("The issue {} is already in the status category that would result from this transition ({}).", issueKey, transitionName);
+                }
+            } else {
+                logger.debug("No transition name was provided so no transition will be performed for this operation: {}.", operation);
+            }
+        } else {
+            logger.debug("No transition required for this issue: {}.", issueKey);
+        }
+        return false;
+    }
+
+    public Optional<TransitionComponent> retrieveIssueTransition(String issueKey, String transitionName) throws IntegrationException {
+        TransitionsResponseModel transitions = issueService.getTransitions(issueKey);
+        return transitions.findFirstTransitionByName(transitionName);
+    }
+
+    private boolean isTransitionRequired(String issueKey, ItemOperation operation) throws IntegrationException {
+        StatusDetailsComponent statusDetailsComponent = issueService.getStatus(issueKey);
+        StatusCategory statusCategory = statusDetailsComponent.getStatusCategory();
+        if (ItemOperation.ADD.equals(operation)) {
+            // Should reopen?
+            return DONE_STATUS_CATEGORY_KEY.equals(statusCategory.getKey());
+        } else if (ItemOperation.DELETE.equals(operation)) {
+            // Should resolve?
+            return TODO_STATUS_CATEGORY_KEY.equals(statusCategory.getKey());
+        }
+        return false;
+    }
+
+    private void performTransition(String issueKey, String transitionName) throws IntegrationException {
+        logger.debug("Attempting the transition '{}' on the issue '{}'", transitionName, issueKey);
+        Optional<TransitionComponent> firstTransitionByName = retrieveIssueTransition(issueKey, transitionName);
+        if (firstTransitionByName.isPresent()) {
+            TransitionComponent issueTransition = firstTransitionByName.get();
+            String transitionId = issueTransition.getId();
```

Comment:
Since we're doing this on something that was an optional, we could call getId before we finish use of the optional then check if the id isPresent.

Reply:
Good call. Originally, we were getting the `StatusCategory` that it would transition to from the `TransitionComponent`, but now the onus of knowing the issue state is on the caller of this method.